### PR TITLE
Implement token resolution and natural language service under mock classifier service

### PR DESCRIPTION
**Pull request by priyankasharma55**
Implementing token service as per https://github.com/buerokratt/Mock-Classifier/issues/14
Added unit test cases

### CODE REVIEW

**Comment by priyankasharma55** [comment]
Code coverage captured for the MinistryClassifierService
image

File: src/MockClassifier.Api/Interfaces/IClassifierService.cs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+﻿namespace MockClassifier.Api.Interfaces
+{
+public interface IClassifierService
```
Comment (martinkearn) [inline src/MockClassifier.Api/Interfaces/IClassifierService.cs:3]: Ideally, interfaces should have XML based comments: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/xmldoc/

File: src/MockClassifier.Api/Services/NaturalLanguageService.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+namespace MockClassifier.Api.Services
+{
+    public class NaturalLanguageService : IClassifierService
+    {
+
+        public string[] Classify(string messageBody)
+        {
+            var token = "";
+            switch (messageBody)
```
Comment (martinkearn) [inline src/MockClassifier.Api/Services/NaturalLanguageService.cs:12]: nit: I may be incorrect here, but would a switch expression make this a little more succinct? https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/switch-expression

Comment (martinkearn) [inline src/MockClassifier.Api/Services/NaturalLanguageService.cs:12]: Would it be safer to cast this to lowercase and change the case comparisons to lower too? It would eliminate people entering things in the wrong case when demonstrating

File: src/MockClassifier.Api/Services/TokenService.cs [hunk 3]
```diff
@@ -0,0 +1,45 @@
+﻿using MockClassifier.Api.Models;
+using MockClassifier.Api.Interfaces;
+using System.Text.RegularExpressions;
+
+
+namespace MockClassifier.Api.Services
+{
+    public class TokenService : IClassifierService
+    {
+       private readonly Regex ministryRegex = new("<random>|<justice>|<education>|<defence>|<environment>|<cultural>|<rural>|<economic>|<finance>|<social>|<interior>|<foreign>", RegexOptions.Compiled);
```
Comment (martinkearn) [inline src/MockClassifier.Api/Services/TokenService.cs:10]: I'm not sure but is there any way this Regex can be derived from the `MinistryEnum` to avoid having them all written out again?

File: src/MockClassifier.Api/Services/TokenService.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+       public string[] Classify(string messageBody)
+       {
+            //search for 11 ministries' label and strip <> and return in lowercase if found
+            var tokenSections = ministryRegex.Matches(messageBody);
```
Comment (martinkearn) [inline src/MockClassifier.Api/Services/TokenService.cs:15]: Do we need to null check `messageBody`?

File: src/MockClassifier.UnitTests/NaturalLanguageServiceTest.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+        }
+
+        [Theory]
+        [InlineData("I want to register my child at school", new string[] { "education" })]
+        [InlineData("I wish to understand what benefits my family are entitled to", new string[] { "social" })]
+        [InlineData("How do I arrange for my COVID-19 booster vaccination", new string[] { "social" })]
+        [InlineData("How do I get to Lahemaa park?", new string[] { "environment" })]
+        [InlineData("I have a question about the Estonian pension system", new string[] { "economic" })]
+        [InlineData("How do I file my annual tax information", new string[] { "economic" })]
+        public void TestClassify(string messageBody, string[] expectedTokens)
```
Comment (martinkearn) [inline src/MockClassifier.UnitTests/NaturalLanguageServiceTest.cs:22]: Can we add a test with an empty string?

File: src/MockClassifier.UnitTests/TokenServiceTest.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        [Theory]
+        [InlineData("<education>", new string[] { "education" })]
+        [InlineData("Please return the <social> minsitry", new string[] { "social" })]
+        [InlineData("I want to see <rural><social> and <environment>", new string[] {"rural","social", "environment" })]
+        [InlineData("<rural> please", new string[] {"rural"}) ]
+        [InlineData("Please return < environment>", new string[] { })]
+        [InlineData("Please return <educationandresearch>", new string[] { })]
+        [InlineData("<Please return <education,social>",new string[] {})]
+
+        public void TestClassify( string messageBody, string[] expectedTokens)
```
Comment (martinkearn) [inline src/MockClassifier.UnitTests/TokenServiceTest.cs:28]: Can we add a test with an empty string?

File: src/MockClassifier.UnitTests/TokenServiceTest.cs [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+using MockClassifier.Api.Models;
+using System;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace MockClassifier.UnitTests
+{
+    public class TokenServiceTest
+    {
+        private readonly TokenService ministryClassifierService;
```
Comment (martinkearn) [inline src/MockClassifier.UnitTests/TokenServiceTest.cs:12]: nit: This is a super minor point, but I normally name the system under test as `sut` so it is clear throughout the tests which system is the one being tested vs mocks and other components. Feel free to ignore, it really does not matter. Discussion here FYI: https://stackoverflow.com/questions/7321407/what-is-sut-and-where-did-it-come-from#:~:text=System%20under%20test%20%28SUT%29%20refers%20to%20a%20system,%20tested%2C%20is%20called%20an%20application%20under%20test.

**Review by martinkearn — APPROVED** [review]
Overall direction is great, just a few minor suggestions but nothing major so I'm happy to approve as-is. If you want to action any of the suggestions, you are welcome to but they are not required.

**Comment by priyankasharma55** [comment]
Thank you @martinkearn for your review comments. will take a look at the changes required

File: src/MockClassifier.Api/Services/NaturalLanguageService.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+            {
+                case "I want to register my child at school": token = Ministry.education.ToString(); break;
+                case "How do I file my annual tax information": token = Ministry.economic.ToString(); break;
+                case "I have a question about the Estonian pension system": token = Ministry.economic.ToString(); break;
+                case "How do I get to Lahemaa park?": token = Ministry.environment.ToString(); break;
+                case "How do I arrange for my COVID-19 booster vaccination": token = Ministry.social.ToString(); break;
+                case "I wish to understand what benefits my family are entitled to": token = Ministry.social.ToString(); break;
+            }
+
+            return new string[] { token };
```
Comment (decodingahmed) [inline src/MockClassifier.Api/Services/NaturalLanguageService.cs:22]: I think this functionality should move into `TokenService` class. We would need this as a fallback in case there are no tokens in the message.

Also, it would save us from injecting multiple `IClassifierService` services into the same controller.

File: src/MockClassifier.Api/Services/TokenService.cs [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                return new string[] { token };
+            }
+            else
+                return tokens.ToArray();
+       }
+
+        private static string GetRandomMinistry()
+        {
+            Array values = Enum.GetValues(typeof(Ministry));
+            Random random = new Random();
```
Comment (decodingahmed) [inline src/MockClassifier.Api/Services/TokenService.cs:39]: Since enum values don't change whilst this application is running, you could probably cache the enum values in the constructor of this class simplify this code.

Please also move `random` to the top and make it readonly:
```
private readonly Random random = new Random();
```

File: src/MockClassifier.UnitTests/NaturalLanguageServiceTest.cs [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+        {
+            var result = naturalLanguageService.Classify(messageBody);
+            Assert.Equal(expectedTokens, result);
+            Assert.NotNull(result);
+        }
```
Comment (decodingahmed) [inline src/MockClassifier.UnitTests/NaturalLanguageServiceTest.cs:27]: Just a reminder to move this test into the other class once you've merged the functionality I mentioned above.

File: src/MockClassifier.Api/Services/TokenService.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
+﻿using MockClassifier.Api.Models;
+using MockClassifier.Api.Interfaces;
+using System.Text.RegularExpressions;
+
+
+namespace MockClassifier.Api.Services
+{
+    public class TokenService : IClassifierService
```
Comment (decodingahmed) [inline src/MockClassifier.Api/Services/TokenService.cs:8]: Please rename this to ClassifierService

File: src/MockClassifier.Api/Interfaces/IClassifierService.cs [hunk 12]
```diff
@@ -0,0 +1,7 @@
...
+{
+    string[] Classify(string messageBody);
+}
+}
```
[the remark this answers is not in the document]
Comment (martinkearn) [reply src/MockClassifier.Api/Interfaces/IClassifierService.cs:7]: This not an action for @priyankasharma55, but regarding our discussion earlier @decodingahmed, this is the sort of thing that StyleCop would sort out rather than us having to look for it in PRs

File: src/MockClassifier.Api/Services/NaturalLanguageService.cs (hunk 8 above)
Comment (martinkearn) [reply src/MockClassifier.Api/Services/NaturalLanguageService.cs:22]: @decodingahmed we've deliberately separated this out into two distinct services; one that deals with tokens and one that deals with natural language. The intention here is on the basis that the natural language bit may grow over time (Cognitive Language Service) even in the context of the mock classifier. Having them separate makes more sense regarding separation of concerns etc. Regarding fall back, if there are no tokens, that is a perfectly normal path and it just means that the token service does not return any tokens (which is fine), we do not need a fall back there.

File: src/MockClassifier.Api/Services/NaturalLanguageService.cs (hunk 2 above)
Comment (priyankasharma55) [reply src/MockClassifier.Api/Services/NaturalLanguageService.cs]: resolved in latest changes

**Comment by priyankasharma55** [comment]
Made most of the changes as suggested

File: src/MockClassifier.Api/Services/TokenService.cs [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+                sb.Append(ministry.ToString());
+                sb.Append(">|");
+            }
+            sb.Append("<random>");
+            ministryRegex = new(sb.ToString(), RegexOptions.Compiled);
+        }
+
+        public string[] Classify(string messageBody)
+        {
+            if(messageBody == null || messageBody.Length == 0)
```
Comment (guypritchard) [inline src/MockClassifier.Api/Services/TokenService.cs:29]: Nit: string.IsNullOrEmpty here?

File: src/MockClassifier.Api/Services/NaturalLanguageService.cs [hunk 14]
```diff
@@ -0,0 +1,29 @@
+﻿using MockClassifier.Api.Models;
+using MockClassifier.Api.Interfaces;
+
+namespace MockClassifier.Api.Services
+{
+    public class NaturalLanguageService : IClassifierService
+    {
+        public string[] Classify(string messageBody)
+        {
+            if (messageBody == null || messageBody.Length == 0)
```
Comment (guypritchard) [inline src/MockClassifier.Api/Services/NaturalLanguageService.cs:10]: Super nit: string.IsNullOrEmpty?

File: src/MockClassifier.Api/Services/TokenService.cs [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            if (tokens.Any() && tokens.Contains("random"))
+            {
+                var token = GetRandomMinistry();
+                return new string[] { token };
+            }
+            else
+                return tokens.ToArray();
+        }
+
+        private string GetRandomMinistry()
```
Comment (guypritchard) [inline src/MockClassifier.Api/Services/TokenService.cs:53]: This again is a super nit - but is this returning the Ministry name (as a string) not the ministry?

**Review by guypritchard — APPROVED** [review]
This is a really nice PR

File: src/MockClassifier.Api/Services/TokenService.cs (hunk 13 above)
Comment (priyankasharma55) [reply src/MockClassifier.Api/Services/TokenService.cs]: Makes sense. done

**Review by decodingahmed — APPROVED** [review]
Looks good :)

**Comment by martinkearn** [comment]
Looks good to merge now @priyankasharma55
